src/mbgl/map/transform.hpp

```diff
@@ -132,22 +132,87 @@ class Transform : private util::noncopyable {
     FreeCameraOptions getFreeCameraOptions() const;
     void setFreeCameraOptions(const FreeCameraOptions& options);
 
+    struct Animation {
+        TimePoint start;
+        Duration duration;
+        AnimationOptions options;
+        bool ran = false;      // Did this property animation run this frame
+        bool finished = false; // Did we execute the finish frame for this property animation this frame
+        bool done = false;     // Did this property animation reach the end of the frame
+        bool panning = false, scaling = false, rotating = false;
+
+        // Anchor
+        std::optional<ScreenCoordinate> anchor;
+        LatLng anchorLatLng;
+
+        Animation(TimePoint start_,
+                  Duration duration_,
+                  AnimationOptions options_,
+                  bool panning_,
+                  bool scaling_,
+                  bool rotating_)
+            : start(start_),
+              duration(duration_),
+              options(options_),
+              panning(panning_),
+              scaling(scaling_),
+              rotating(rotating_) {}
+
+        double interpolant(TimePoint);
+
+        bool isAnimated() const { return duration != Duration::zero(); }
+    };
+
 private:
+    template <class T>
+    struct Property {
+        std::shared_ptr<Animation> animation;
```

Comment:
Why is this a shared pointer?

Reply:
The `Animation` object is used to keep track of the properties current animation. It keeps track whether we've ran the animation's frame callback, animation's finish callback, and the panning, scaling, and rotating state.

So, to avoid updating the above for every property we use a shared pointer to ensure that if any one of the properties finished running the animation's frame or finish callbacks then the other one doesn't do it again.